File: flutter/invisibar.dart
/// Invisibar — https://github.com/gokhunguneyhan/invisibar
///
/// Hide the iOS status bar for your product screenshots and recordings, which also
/// removes the red screen-recording indicator from the Dynamic Island.
///
/// No plugins, no dependencies beyond Flutter itself.
///
///   // 1. wrap your app root
///   runApp(const Invisibar(child: MyApp()));
///
///   // 2. anywhere you can spare a footnote, e.g. under a settings list
///   const InvisibarLink()
///
/// Tap the footnote, pick Hide or Replace, then scroll the link out of frame and
/// record.
///
/// RELEASE SAFETY. Everything is behind `kDebugMode`, which is a compile-time
/// constant in release builds, so the tree-shaker removes the dead branch.
/// `InvisibarLink` builds a `SizedBox.shrink()` and `Invisibar` returns its child
/// untouched. Verify on your own release build, and use a control string or a clean
/// zero proves nothing:
///
///   flutter build ios --release
///   strings -a build/ios/.../Runner.app/Frameworks/App.framework/App \
///     | grep -c Invisibar
///
/// iOS only by nature: Android has no Dynamic Island and draws its own recording
/// indicator, which this cannot touch.
///
/// MIT licensed.

import 'package:flutter/foundation.dart' show kDebugMode;
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

enum InvisibarMode { off, hide, replace }

/// Deliberately not persisted. A capture mode that resets on restart is the safer
/// default: you cannot get stuck with a hidden status bar and no way back.
final ValueNotifier<InvisibarMode> _mode =
    ValueNotifier<InvisibarMode>(InvisibarMode.off);

/// Wrap your app root. Returns the child untouched in release.
class Invisibar extends StatelessWidget {
  const Invisibar({super.key, required this.child});

  final Widget child;

  @override
  Widget build(BuildContext context) {
    if (!kDebugMode) return child;
    return ValueListenableBuilder<InvisibarMode>(
      valueListenable: _mode,
      builder: (context, mode, _) {
        SystemChrome.setEnabledSystemUIMode(
          SystemUiMode.manual,
          overlays: mode == InvisibarMode.off
              ? SystemUiOverlay.values
              : const <SystemUiOverlay>[],
        );
        return Stack(
          children: [
            child,
            if (mode == InvisibarMode.replace)
              const Positioned(
                top: 0,
                left: 0,
                right: 0,
                child: IgnorePointer(child: _InvisibarStatusBar()),
              ),
          ],
        );
      },
    );
  }
}

/// A dim footnote reading "Invisibar". Builds nothing in release.
class InvisibarLink extends StatelessWidget {
  const InvisibarLink({super.key});

  @override
  Widget build(BuildContext context) {
    if (!kDebugMode) return const SizedBox.shrink();
    return ValueListenableBuilder<InvisibarMode>(
      valueListenable: _mode,
      builder: (context, mode, _) {
        final dim = Theme.of(context).textTheme.bodySmall?.color;
        return GestureDetector(
          onTap: () => showModalBottomSheet<void>(
            context: context,
            showDragHandle: true,
            builder: (_) => const _InvisibarSheet(),
          ),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              Text('Invisibar', style: TextStyle(fontSize: 13, color: dim)),
              if (mode != InvisibarMode.off) ...[
                const SizedBox(width: 5),
                Container(
                  width: 5,
                  height: 5,
                  decoration: BoxDecoration(color: dim, shape: BoxShape.circle),
                ),
              ],
            ],
          ),
        );
      },
    );
  }
}

class _InvisibarSheet extends StatelessWidget {
  const _InvisibarSheet();

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final dim = theme.textTheme.bodySmall?.color;

    Widget row(InvisibarMode m, String title, String subtitle) {
      return ValueListenableBuilder<InvisibarMode>(
        valueListenable: _mode,
        builder: (context, mode, _) => InkWell(
          // Deliberately does not pop: you often want to try one, look, and try
          // the other.
          onTap: () => _mode.value = m,
          child: Padding(
            padding: const EdgeInsets.symmetric(vertical: 12),
            child: Row(
              children: [
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(title, style: const TextStyle(fontSize: 17)),
                      const SizedBox(height: 2),
                      Text(subtitle,
                          style: TextStyle(fontSize: 12, color: dim)),
                    ],
                  ),
                ),
                Icon(
                  mode == m ? Icons.check_circle : Icons.circle_outlined,
                  color: mode == m ? theme.colorScheme.primary : dim,
                  size: 22,
                ),
              ],
            ),
          ),
        ),
      );
    }

    return Padding(
      padding: const EdgeInsets.fromLTRB(24, 0, 24, 28),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Text('Invisibar',
              style: theme.textTheme.titleLarge
                  ?.copyWith(fontWeight: FontWeight.w600)),
          const SizedBox(height: 6),
          Text(
            'For screenshots and screen recordings.\n'
            'Also removes the red recording indicator.',
            textAlign: TextAlign.center,
            style: TextStyle(fontSize: 15, color: dim),
          ),
          const SizedBox(height: 20),
          row(InvisibarMode.off, 'Off', 'Leave the real status bar alone.'),
          const Divider(height: 1),
          row(InvisibarMode.hide, 'Hide status bar',
              'No clock, battery, signal or recording indicator.'),
          const Divider(height: 1),
          row(InvisibarMode.replace, 'Replace status bar',
              'Draws a clean 9:41 and a full battery instead.'),
          const SizedBox(height: 20),
          Text('Built by Gokhun Guneyhan',
              style: TextStyle(fontSize: 13, color: dim)),
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              _LinkText('x', 'https://x.com/gokhunguneyhan', dim),
              Text('/', style: TextStyle(fontSize: 16, color: dim)),
              _LinkText('github',
                  'https://github.com/gokhunguneyhan/invisibar', dim),
            ],
          ),
        ],
      ),
    );
  }
}

class _LinkText extends StatelessWidget {
  const _LinkText(this.label, this.url, this.color);

  final String label;
  final String url;
  final Color? color;

  @override
  Widget build(BuildContext context) {
    // No url_launcher dependency: copying the URL is enough for a dev tool, and
    // keeping this file plugin-free is worth more than opening a browser.
    return TextButton(
      onPressed: () => Clipboard.setData(ClipboardData(text: url)),
      child: Text(label, style: TextStyle(fontSize: 16, color: color)),
    );
  }
}

/// Geometry measured from 1206x2622 captures of a 402pt-wide 3x device on iOS 26,
/// expressed as fractions of the width so it travels to other sizes. Check it
/// against a screenshot of your own device before shooting a lot of footage.
///
/// Do NOT calibrate against a capture taken while screen recording: the Dynamic
/// Island is expanded then and pushes the time left, 61.5pt versus 74.2pt idle. The
/// vertical band is the same in both, which is why it is the number to trust.
///
/// No Wi-Fi glyph: arcs need a custom painter, and this stays deliberately small.
class _InvisibarStatusBar extends StatelessWidget {
  const _InvisibarStatusBar({this.time = '9:41', this.batteryLevel = 1.0});

  final String time;
  final double batteryLevel;

  @override
  Widget build(BuildContext context) {
    final width = MediaQuery.sizeOf(context).width;
    final dark = Theme.of(context).brightness == Brightness.dark;
    final tint = dark ? Colors.white : Colors.black;

    final timeCentre = width * (74.2 / 402);
    final trailing = width * (31.3 / 402);
    const bandTop = 26.7;

    return SizedBox(
      height: 54,
      child: Stack(
        children: [
          Positioned(
            top: bandTop - 1,
            left: 0,
            // A fixed-width box starting at 0 centres the text on its midpoint,
            // so this holds however wide the string renders.
            width: timeCentre * 2,
            child: Text(
              time,
              textAlign: TextAlign.center,
              style: TextStyle(
                  fontSize: 17, fontWeight: FontWeight.w600, color: tint),
            ),
          ),
          Positioned(
            top: bandTop,
            right: trailing,
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.end,
              children: [
                for (final h in const [4.0, 6.3, 8.7, 11.3])
                  Container(
                    width: 3,
                    height: h,
                    margin: const EdgeInsets.only(right: 1.7),
                    decoration: BoxDecoration(
                      color: tint,
                      borderRadius: BorderRadius.circular(0.8),
                    ),
                  ),
                const SizedBox(width: 4.2),
                _Battery(level: batteryLevel, tint: tint),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class _Battery extends StatelessWidget {
  const _Battery({required this.level, required this.tint});

  final double level;
  final Color tint;

  @override
  Widget build(BuildContext context) {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        SizedBox(
          width: 25,
          height: 12,
          child: Stack(
            children: [
              Container(
                decoration: BoxDecoration(
                  border: Border.all(color: tint.withOpacity(0.38), width: 1),
                  borderRadius: BorderRadius.circular(3.8),
                ),
              ),
              Positioned(
                left: 2,
                top: 2,
                child: Container(
                  width: 21 * level.clamp(0.0, 1.0),
                  height: 8,
                  decoration: BoxDecoration(
                    color: tint,
                    borderRadius: BorderRadius.circular(1.8),
                  ),
                ),
              ),
            ],
          ),
        ),
        const SizedBox(width: 1),
        Container(
          width: 1.6,
          height: 4.2,
          decoration: BoxDecoration(
            color: tint.withOpacity(0.38),
            borderRadius: BorderRadius.circular(0.8),
          ),
        ),
      ],
    );
  }
}

File: Invisibar.swift
//  Invisibar.swift
//  https://github.com/gokhunguneyhan/invisibar
//
//  Hide the iOS status bar for your product screenshots and recordings — which also
//  removes the red screen-recording indicator from the Dynamic Island.
//
//  Drop this file into your app. Two call sites:
//
//      @main struct MyApp: App {
//          var body: some Scene {
//              WindowGroup { RootView().invisibar() }    // 1. at the root
//          }
//      }
//
//      // 2. anywhere you can spare a footnote, e.g. under a settings list
//      InvisibarLink()
//
//  Tap the footnote, pick Hide or Replace, then scroll the link out of frame and
//  record. Tap it again to turn it off.
//
//  RELEASE SAFETY. Everything below is a no-op in Release: `InvisibarLink` renders
//  `EmptyView`, `.invisibar()` returns the view untouched, and no UserDefaults key is
//  ever read or written. Both call sites still compile, so you can leave them in.
//
//  Measured on a real Release build, against a control string:
//
//      invisibarMode          Debug 1   Release 0
//      "Hide status bar"      Debug 1   Release 0
//      "9:41"                 Debug 2   Release 0
//      InvisibarLink          Debug 5   Release 2   <- type metadata only
//      didRepairPermissions   Debug 1   Release 2   <- control, must be non-zero
//
//  The type name survives because the API has to exist for your call sites to
//  compile. It carries no behaviour and touches no storage. Verify your own build
//  the same way; see references/verify.md, and use a control string or a clean zero
//  proves nothing at all.
//
//  MIT licensed.

import SwiftUI

// MARK: - Public API

public extension View {
    /// Apply at the OUTERMOST view your app shows. If your root `body` returns early
    /// for any branch — a launch argument, a loading state, a separate onboarding
    /// root — apply this above the branch, or those screens keep their status bar.
    func invisibar() -> some View {
        #if DEBUG
        modifier(InvisibarModifier())
        #else
        self
        #endif
    }
}

/// A dim footnote button reading "Invisibar". Sized to disappear into the bottom of a
/// settings screen, so a small scroll leaves it out of shot.
public struct InvisibarLink: View {
    public init() {}

    public var body: some View {
        #if DEBUG
        InvisibarLinkBody()
        #else
        EmptyView()
        #endif
    }
}

#if DEBUG

// MARK: - Mode

enum InvisibarMode: String {
    case off, hide, replace

    static let storageKey = "invisibarMode"
}

// MARK: - The footnote

private struct InvisibarLinkBody: View {
    @AppStorage(InvisibarMode.storageKey) private var raw = InvisibarMode.off.rawValue
    @State private var sheet = false

    private var mode: InvisibarMode { InvisibarMode(rawValue: raw) ?? .off }

    var body: some View {
        Button { sheet = true } label: {
            HStack(spacing: 5) {
                Text("Invisibar")
                if mode != .off {
                    // Only marking when it is ON, so the resting state stays quiet.
                    Circle().frame(width: 5, height: 5)
                }
            }
            .font(.footnote)
            .foregroundStyle(.secondary)
        }
        .buttonStyle(.plain)
        .sheet(isPresented: $sheet) { InvisibarSheet(raw: $raw) }
    }
}

// MARK: - The sheet

private struct InvisibarSheet: View {
    @Binding var raw: String
    @Environment(\.dismiss) private var dismiss

    private var mode: InvisibarMode { InvisibarMode(rawValue: raw) ?? .off }

    /// Tapping the active row turns it off, so two rows cover three states and there
    /// is no "Off" row to explain.
    private func pick(_ m: InvisibarMode) {
        raw = (mode == m ? .off : m).rawValue
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text("Invisibar")
                .font(.title3.weight(.semibold))
                .padding(.top, 24)

            Text("For screenshots and screen recordings. Also removes the red "
                 + "recording indicator.")
                .font(.subheadline)
                .foregroundStyle(.secondary)
                .fixedSize(horizontal: false, vertical: true)
                .padding(.top, 6)

            VStack(spacing: 0) {
                row(.hide, "Hide status bar",
                    "No clock, battery, signal or recording indicator.")
                Divider().padding(.leading, 2)
                row(.replace, "Replace status bar",
                    "Draws a clean 9:41 and a full battery instead.")
            }
            .padding(.top, 22)

            Spacer(minLength: 24)

            HStack(spacing: 14) {
                Link("x", destination: URL(string: "https://x.com/gokhunguneyhan")!)
                Text("/").foregroundStyle(.tertiary)
                Link("github", destination:
                        URL(string: "https://github.com/gokhunguneyhan/invisibar")!)
            }
            .font(.footnote)
            .foregroundStyle(.secondary)
            .padding(.bottom, 20)
        }
        .padding(.horizontal, 24)
        .frame(maxWidth: .infinity, alignment: .leading)
        .presentationDetents([.height(340)])
        .presentationDragIndicator(.visible)
    }

    private func row(_ m: InvisibarMode, _ title: String, _ subtitle: String) -> some View {
        Button {
            pick(m)
            // Close on selecting, so the sheet is not in the way of the thing you
            // are about to record. Turning one OFF keeps it open.
            if mode == m { dismiss() }
        } label: {
            HStack(spacing: 12) {
                VStack(alignment: .leading, spacing: 2) {
                    Text(title).font(.body)
                    Text(subtitle).font(.caption).foregroundStyle(.secondary)
                        .fixedSize(horizontal: false, vertical: true)
                }
                Spacer(minLength: 8)
                Image(systemName: mode == m ? "checkmark.circle.fill" : "circle")
                    .foregroundStyle(mode == m ? Color.accentColor : Color.secondary)
                    .font(.system(size: 20))
            }
            .contentShape(Rectangle())
            .padding(.vertical, 12)
        }
        .buttonStyle(.plain)
    }
}

// MARK: - The modifier

private struct InvisibarModifier: ViewModifier {
    @AppStorage(InvisibarMode.storageKey) private var raw = InvisibarMode.off.rawValue

    private var mode: InvisibarMode { InvisibarMode(rawValue: raw) ?? .off }

    func body(content: Content) -> some View {
        content
            .statusBarHidden(mode != .off)
            .overlay(alignment: .top) {
                if mode == .replace {
                    InvisibarStatusBar().allowsHitTesting(false)
                }
            }
    }
}

// MARK: - The drawn status bar

/// A stand-in for the real status bar, for `replace` mode.
///
/// Geometry measured from 1206x2622 captures of a 402pt-wide 3x device on iOS 26, and
/// expressed as fractions of the width so it travels to other sizes. Not pixel-exact
/// everywhere: check it against a screenshot of your own device before shooting a lot
/// of footage.
///
/// Do not calibrate against a capture taken WHILE screen recording. The Dynamic Island
/// is expanded in that state and pushes the time left — 61.5pt versus 74.2pt idle. The
/// vertical band is the same in both, which is why it is the number to trust.
struct InvisibarStatusBar: View {
    var time: String = "9:41"
    var batteryLevel: Double = 1.0
    var showsCellular: Bool = true
    var showsWiFi: Bool = true

    @Environment(\.colorScheme) private var scheme

    // Fractions of the screen width, from the reference capture.
    private let timeCentreFraction = 74.2 / 402.0
    private let trailingFraction = 31.3 / 402.0
    private let bandTop: CGFloat = 26.7      // identical in every capture taken

    private var tint: Color { scheme == .dark ? .white : .black }

    var body: some View {
        GeometryReader { geo in
            let w = geo.size.width
            ZStack(alignment: .topLeading) {
                Text(time)
                    .font(.system(size: 17, weight: .semibold))
                    // A fixed-width frame centred on 0 puts the midpoint at half the
                    // width, so this places the time regardless of how wide it renders.
                    .frame(width: w * timeCentreFraction * 2)
                    .position(x: w * timeCentreFraction, y: bandTop + 5.8)

                HStack(spacing: 4.2) {
                    if showsCellular {
                        Image(systemName: "cellularbars")
                            .font(.system(size: 15, weight: .semibold))
                    }
                    if showsWiFi {
                        Image(systemName: "wifi")
                            .font(.system(size: 15, weight: .semibold))
                    }
                    BatteryGlyph(level: batteryLevel, tint: tint)
                }
                .frame(maxWidth: .infinity, alignment: .trailing)
                .padding(.trailing, w * trailingFraction)
                .position(x: w / 2, y: bandTop + 5.8)
            }
            .foregroundStyle(tint)
        }
        .frame(height: 54)
        .ignoresSafeArea(edges: .top)
    }
}

/// 25 x 12pt body with a nub, matching Apple's proportions. The outline and nub are
/// translucent; the level fill is solid, as iOS draws it.
private struct BatteryGlyph: View {
    var level: Double
    var tint: Color

    var body: some View {
        HStack(spacing: 1.0) {
            ZStack(alignment: .leading) {
                RoundedRectangle(cornerRadius: 3.8, style: .continuous)
                    .stroke(tint.opacity(0.38), lineWidth: 1)
                    .frame(width: 25, height: 12)
                RoundedRectangle(cornerRadius: 1.8, style: .continuous)
                    .fill(tint)
                    .frame(width: max(0, 21 * min(max(level, 0), 1)), height: 8)
                    .padding(.leading, 2)
            }
            Capsule().fill(tint.opacity(0.38)).frame(width: 1.6, height: 4.2)
        }
    }
}

#endif
